kombu/utils/objects.py

```diff
@@ -25,10 +27,17 @@ def __init__(self, fget=None, fset=None, fdel=None):
             # This is a backport so we set this ourselves.
             self.attrname = self.func.__name__
 
+        if not hasattr(self, 'lock'):
+            # Prior to Python 3.12, functools.cached_property has an
+            # undocumented lock which is required for thread-safe __set__
+            # and __delete__. Create one if it isn't already present.
+            self.lock = RLock()
+
     def __get__(self, instance, owner=None):
         # TODO: Remove this after we drop support for Python<3.8
         #  or fix the signature in the cached_property package
-        return super().__get__(instance, owner)
+        with self.lock:
+            return super().__get__(instance, owner)
```

Comment:
Doesn't this now prevent concurrent reads of the cached value? I think it's only necessary to lock during writes, isn't it?

Reply:
Reads need to synchronize with writes, otherwise a write could be lost if a read found no value present but the write occurred before the read computed and stored the new value. If the presence check and computation were here rather in the superclass only locking after the initial read might be valid, but what I've written here matches the standard library behavior before Python 3.12.

Reply:
ok, makes sense, especially if stdlib did the same